False-positive for W0706 in the case of non-trivial exception inheritance

If you want to trap a parent exception, but skip certain child exceptions, PyLint sometimes emits "W0706: The except handler raises immediately (try-except-raise)" warnings. In the simplest case where the code raises a direct child but traps the parent, PyLint is happy. Complicating that (trap multiple children, more than one level of inheritance) seems to make things go awry.

Maybe I should take a stab at fixing this; I'm assuming the framework is there because the inheritance is checked on one level, just need to add a few more cases...

### Steps to reproduce
```
class ParentException(Exception):
    pass

class ChildException(ParentException):
    pass

class SiblingException(ParentException):
    pass

class GrandchildException(ChildException):
    pass

def two_levels():
    try:
        pass
    except GrandchildException:
        raise  # W0706: The except handler raises immediately
    except ParentException:
        pass

def no_parent_all_children():
    try:
        pass
    except (ChildException, SiblingException):
        raise  # W0706
    except ParentException:
        pass

def explicit_multilevel():
    """A bit contrived"""
    try:
        pass
    except GrandchildException:
        raise  # W0706
    except (ChildException, ParentException):
        pass

def just_the_leaves():
    """Maybe more practical"""
    try:
        pass
    except (GrandchildException, SiblingException):
        raise  # W0706
    except (ChildException, ParentException):
        pass

def leaves_w_implicit_child():
    try:
        pass
    except (GrandchildException, SiblingException):
        raise  # W0706
    except ParentException:
        pass

def lintsok():
    try:
        pass
    except ChildException:
        raise
    except ParentException:
        pass

def lintsok2():
    try:
        pass
    except GrandchildException:
        raise
    except ChildException:
        pass
```


### Current behavior
```
$ pylint immediateraise.py                    
************* Module immediateraise                                                      
immediateraise.py:16:4: W0706: The except handler raises immediately (try-except-raise)  
immediateraise.py:24:4: W0706: The except handler raises immediately (try-except-raise)  
immediateraise.py:33:4: W0706: The except handler raises immediately (try-except-raise)  
immediateraise.py:42:4: W0706: The except handler raises immediately (try-except-raise)  
immediateraise.py:50:4: W0706: The except handler raises immediately (try-except-raise)  

------------------------------------------------------------------ 
Your code has been rated at 9.12/10 (previous run: 9.12/10, +0.00) 
```

### Expected behavior
No warnings.

### pylint --version output
```
pylint 2.1.1                                  
astroid 2.0.4                                 
Python 3.6.4 (default, Mar 26 2018, 16:21:07) 
[GCC 5.4.0 20160609]                          
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -15,6 +15,10 @@ Release date: TBA
 
      Close #2513
 
+   * `try-except-raise` checker now handles multilevel inheritance hirerachy for exceptions correctly.
+
+     Close #2484
+
    * Add a new check, ``simplifiable-if-expression`` for expressions like ``True if cond else False``.
 
      Close #2487

--- a/doc/whatsnew/2.2.rst
+++ b/doc/whatsnew/2.2.rst
@@ -20,6 +20,10 @@ New checkers
 Other Changes
 =============
 
+* `try-except-raise` checker now handles multilevel inheritance hirerachy for exceptions correctly.
+
+  Close #2484
+
 * Ignore import x.y.z as z cases for checker `useless-import-alias`.
 
 * `unnecessary-pass` is now also emitted when a function or class contains only docstring and pass statement, 

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -1135,10 +1135,7 @@ def is_subclass_of(child: astroid.ClassDef, parent: astroid.ClassDef) -> bool:
     if not all(isinstance(node, astroid.ClassDef) for node in (child, parent)):
         return False
 
-    for base in child.ancestors():
-        if parent.name == base.name:
+    for ancestor in child.ancestors():
+        if astroid.helpers.is_subtype(ancestor, parent):
             return True
-        # recursively check child's parent for actual parent
-        child = safe_infer(base)
-        is_subclass_of(child, parent)
     return False

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -1125,17 +1125,22 @@ def is_postponed_evaluation_enabled(node: astroid.node_classes.NodeNG) -> bool:
     )
 
 
-def is_subclass_of(node_a: astroid.ClassDef, node_b: astroid.ClassDef) -> bool:
+def is_subclass_of(child: astroid.ClassDef,
+                   parent: astroid.ClassDef) -> bool:
     """
     Check if first node is a subclass of second node.
-    :param node_a: Node to check for subclass.
-    :type node_a: astroid.ClassDef
-    :param node_b: Node to check for superclass.
-    :type node_b: astroid.ClassDef
-    :returns: True if node_a is derived from node_b. False otherwise.
-    :rtype: bool
+    :param child: Node to check for subclass.
+    :param parent: Node to check for superclass.
+    :returns: True if child is derived from parent. False otherwise.
     """
-    if not all(isinstance(node, astroid.ClassDef) for node in (node_a, node_b)):
+    if not all(isinstance(node, astroid.ClassDef) for node in (child, parent)):
         return False
 
-    return node_b.name in {base.name for base in node_a.bases}
+    
+    for base in child.ancestors():
+        if parent.name == base.name:
+            return True
+        # recursively check child's parent for actual parent
+        child = safe_infer(base)
+        is_subclass_of(child, parent)
+    return False

--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -1125,8 +1125,7 @@ def is_postponed_evaluation_enabled(node: astroid.node_classes.NodeNG) -> bool:
     )
 
 
-def is_subclass_of(child: astroid.ClassDef,
-                   parent: astroid.ClassDef) -> bool:
+def is_subclass_of(child: astroid.ClassDef, parent: astroid.ClassDef) -> bool:
     """
     Check if first node is a subclass of second node.
     :param child: Node to check for subclass.
@@ -1136,7 +1135,6 @@ def is_subclass_of(child: astroid.ClassDef,
     if not all(isinstance(node, astroid.ClassDef) for node in (child, parent)):
         return False
 
-    
     for base in child.ancestors():
         if parent.name == base.name:
             return True

--- a/pylint/test/functional/try_except_raise.py
+++ b/pylint/test/functional/try_except_raise.py
@@ -1,5 +1,5 @@
 # pylint:disable=missing-docstring, unreachable, bad-except-order, bare-except, unnecessary-pass
-# pylint: disable=undefined-variable
+# pylint: disable=undefined-variable, broad-except
 try:
     int("9a")
 except:  # [try-except-raise]
@@ -103,3 +103,18 @@ except invalid_name:  # [try-except-raise]
     raise
 except TypeError:
     pass
+
+try:
+    pass
+except (FileNotFoundError, PermissionError):
+    raise
+except Exception:
+    print("a failure")
+
+try:
+    pass
+except (FileNotFoundError, PermissionError):
+    raise
+except (Exception,):
+    print("a failure")
+    

